File: src/utils/gameLogic/physics.ts
import { v7 } from 'uuid';

import {
  BOARD_WIDTH,
  BOARD_HEIGHT,
  TIMER_CONFIG,
} from '@/constants/gameConfig';
import type {
  GameBoard,
  Square,
  FallingColumn,
  FallingCell,
} from '@/types/game';

import { copyBoard, coordToString } from './helpers';

/**
 * Create falling cells from board
 */
export function createFallingColumns(
  board: GameBoard,
  fallingColumns: FallingColumn[]
): { newFallingColumns: FallingColumn[]; newBoard: GameBoard } {
  const newBoard = copyBoard(board);
  const newFallingColumns: FallingColumn[] = [];

  // Process each column
  for (let x = 0; x < BOARD_WIDTH; x++) {
    // Find any cells that need to fall in this column
    const fallingCells = findFallingCellsInColumn(newBoard, x);

    // Check for existing falling cells in this column
    const existingColumn = fallingColumns.find(col => col.x === x);

    // Include column if it has new falling cells OR existing falling cells
    if (fallingCells.length > 0 || existingColumn) {
      // Merge with any existing falling cells in this column
      const mergedCells = existingColumn
        ? [...existingColumn.cells, ...fallingCells]
        : fallingCells;

      // Create new column with merged cells sorted by y position
      newFallingColumns.push({
        x,
        cells: mergedCells.sort((a, b) => b.y - a.y), // Sort bottom to top
        timer: existingColumn?.timer ?? 0,
      });
    }
  }

  return { newFallingColumns, newBoard };
}

/**
 * Find falling cells in a column
 */
export function findFallingCellsInColumn(
  board: GameBoard,
  x: number
): FallingCell[] {
  const cells: FallingCell[] = [];
  let foundGap = false;

  for (let y = BOARD_HEIGHT - 1; y >= 0; y--) {
    const cell = board[y][x];

    if (cell === 0) {
      foundGap = true;
    } else if (foundGap) {
      cells.push({
        id: v7(),
        y,
        color: cell,
      });
      board[y][x] = 0;
    }
  }

  return cells;
}

/**
 * Update falling columns physics
 */
export function updateFallingColumns(
  board: GameBoard,
  fallingColumns: FallingColumn[]
): { newBoard: GameBoard; newFallingColumns: FallingColumn[] } {
  const newBoard = copyBoard(board);
  const newFallingColumns: FallingColumn[] = [];

  for (const column of fallingColumns) {
    const newTimer = column.timer + 1;
    const isTimerReached = newTimer >= TIMER_CONFIG.FALLING_CELL_INTERVAL;
    const updatedCells: FallingCell[] = [];

    // Process each cell in the column
    for (const cell of column.cells) {
      const nextY = cell.y + 1;

      if (nextY >= BOARD_HEIGHT || newBoard[nextY][column.x] !== 0) {
        newBoard[cell.y][column.x] = cell.color;
        continue;
      }

      updatedCells.push({
        id: cell.id,
        y: isTimerReached ? nextY : cell.y,
        color: cell.color,
      });
    }

    // Only keep columns that still have falling cells
    if (updatedCells.length > 0) {
      newFallingColumns.push({
        x: column.x,
        cells: updatedCells,
        timer: isTimerReached ? 0 : newTimer,
      });
    }
  }

  return { newBoard, newFallingColumns };
}

/**
 * Clear marked cells from the board and apply gravity
 * Returns the new board with marked cells cleared and blocks settled
 */
export function clearMarkedCellsAndApplyGravity(
  board: GameBoard,
  markedCells: Square[],
  fallingColumns: FallingColumn[]
): { newBoard: GameBoard; newFallingColumns: FallingColumn[] } {
  // Create a copy of the board to avoid mutation
  const newBoard = copyBoard(board);

  // Create a set of positions to clear for efficient lookup
  const cellsToClear = new Set<string>();
  for (const cell of markedCells) {
    cellsToClear.add(coordToString(cell.x, cell.y));
  }

  // Clear all marked cells
  for (let y = 0; y < BOARD_HEIGHT; y++) {
    for (let x = 0; x < BOARD_WIDTH; x++) {
      if (cellsToClear.has(coordToString(x, y))) {
        newBoard[y][x] = 0;
      }
    }
  }

  return createFallingColumns(newBoard, fallingColumns);
}

File: python/game/physics.py
"""
physics.py — Falling cell physics.
Port of src/utils/gameLogic/physics.ts
"""

import uuid
from typing import List, Tuple
from .constants import BOARD_WIDTH, BOARD_HEIGHT, FALLING_CELL_INTERVAL
from .types import FallingCell, FallingColumn, Square


def _new_cell_id() -> str:
    """Generate a UUID for a falling cell."""
    return str(uuid.uuid4())


def find_falling_cells_in_column(
    board: List[List[int]],
    x: int,
) -> List[FallingCell]:
    """
    Find cells in column x that should start falling (cells above a gap).
    Mutates board in-place: clears the found cells from the board.
    Returns cells sorted bottom-to-top (highest y first).
    """
    cells: List[FallingCell] = []
    found_gap = False

    for y in range(BOARD_HEIGHT - 1, -1, -1):
        cell = board[y][x]

        if cell == 0:
            found_gap = True
        elif found_gap:
            cells.append(FallingCell(id=_new_cell_id(), y=y, color=cell))
            board[y][x] = 0

    return cells


def create_falling_columns(
    board: List[List[int]],
    falling_columns: List[FallingColumn],
) -> Tuple[List[FallingColumn], List[List[int]]]:
    """
    Identify columns with unsupported (floating) cells and create
    FallingColumn records for them.

    Returns (new_falling_columns, new_board) where cells-to-fall have
    been removed from new_board.
    """
    new_board = [row[:] for row in board]
    new_falling_columns: List[FallingColumn] = []

    for x in range(BOARD_WIDTH):
        falling_cells = find_falling_cells_in_column(new_board, x)
        existing_col = next((c for c in falling_columns if c.x == x), None)

        if falling_cells or existing_col:
            merged = (existing_col.cells if existing_col else []) + falling_cells
            # Sort bottom-to-top: highest y value first
            merged.sort(key=lambda c: c.y, reverse=True)
            new_falling_columns.append(FallingColumn(
                x=x,
                cells=merged,
                timer=existing_col.timer if existing_col else 0,
            ))

    return new_falling_columns, new_board


def update_falling_columns(
    board: List[List[int]],
    falling_columns: List[FallingColumn],
) -> Tuple[List[List[int]], List[FallingColumn]]:
    """
    Advance falling cells by one tick.
    Cells that have landed are written to the board.
    Returns (new_board, new_falling_columns).
    """
    new_board = [row[:] for row in board]
    new_falling_columns: List[FallingColumn] = []

    for column in falling_columns:
        new_timer = column.timer + 1
        is_timer_reached = new_timer >= FALLING_CELL_INTERVAL
        updated_cells: List[FallingCell] = []

        for cell in column.cells:
            next_y = cell.y + 1

            if next_y >= BOARD_HEIGHT or new_board[next_y][column.x] != 0:
                # Cell has landed — write to board
                new_board[cell.y][column.x] = cell.color
                continue

            updated_cells.append(FallingCell(
                id=cell.id,
                y=next_y if is_timer_reached else cell.y,
                color=cell.color,
            ))

        if updated_cells:
            new_falling_columns.append(FallingColumn(
                x=column.x,
                cells=updated_cells,
                timer=0 if is_timer_reached else new_timer,
            ))

    return new_board, new_falling_columns


def clear_marked_cells_and_apply_gravity(
    board: List[List[int]],
    marked_cells: List[Square],
    falling_columns: List[FallingColumn],
) -> Tuple[List[List[int]], List[FallingColumn]]:
    """
    Clear all marked cells from the board, then re-detect falling columns.
    Returns (new_board, new_falling_columns).
    """
    new_board = [row[:] for row in board]

    # Build set of positions to clear
    cells_to_clear = {(cell.x, cell.y) for cell in marked_cells}

    for y in range(BOARD_HEIGHT):
        for x in range(BOARD_WIDTH):
            if (x, y) in cells_to_clear:
                new_board[y][x] = 0

    return create_falling_columns(new_board, falling_columns)
